### PR TITLE
Piecewise encode and decode

### PR DESCRIPTION
**Pull request by jemrobinson**
# Piecewise encoder/decoder
- Encode by "unfolding" the input into overlapping patches
- Decode by "folding" the input into a larger whole
- Configurable number of convolutional blocks to add non-linearity
- Configurable final layer to force output into [0, 1]

# Interim conclusions
**You need at least one ConvBlock on the encoder side or the loss plateaus**
no-encoder-cnn-vs-at-least-one-block

**Number of decoder ConvBlocks seems to be more correlated with loss than number of encoder ConvBlocks**
correlations

# Comparisons to other encoder/decoder pairs

**CNN** (9.15it/s)
cnn-unet-cnn

**Naive** (30.87it/s)
naive-unet-naive

**Piecewise** (4.79it/s)
piecewise-unet-piecewise

# Notes
- :warning: Encoder names are changed from `encoder_0` to `encoder_era5` etc. This means that checkpoints trained before this PR might not be loadable after this is merged.

### CODE REVIEW

File: icenet_mp/models/decoders/piecewise_decoder.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        # Construct list of layers
+        layers: list[nn.Module] = []
+
+        # Add a convolutional block to get the required number of channels
+        layers.append(
+            CommonConvBlock(
+                self.data_space_in.channels,
+                input_channels_required,
+                kernel_size=3,
+                activation="SiLU",
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/decoders/piecewise_decoder.py]: Done in 35b380fe

File: icenet_mp/models/encoders/piecewise_encoder.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                stride=strides,
+                padding=strides,
+            ),
+            # Permute dimensions: [N, n_patches, C * patch_area]
+            Permute((0, 2, 1)),
+            # Unflatten the patches: [N, n_patches, C, patch_h, patch_w]
+            nn.Unflatten(2, (-1, *self.data_space_out.shape)),
+            # Combine into a single channel dimension: [N, C * n_patches, patch_h, patch_w]
+            nn.Flatten(1, 2),
+        ]
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/encoders/piecewise_encoder.py]: That's only a requirement if you need the output to exactly reproduce the input i.e. just for this test.
